src/tree/iterators.rs

```diff
@@ -0,0 +1,401 @@
+}
+
+///
+/// An Iterator over the children of a `Node`.
+///
+/// Iterates over `NodeId`s instead of over the `Node`s themselves.
+///
+pub struct ChildrenIds<'a> {
+    child_ids: Iter<'a, NodeId>,
+}
+
+impl<'a, T> IteratorNew<'a, T, ChildrenIds<'a>> for ChildrenIds<'a> {
+    fn new(tree: &'a Tree<T>, node_id: NodeId) -> ChildrenIds<'a> {
+        ChildrenIds { child_ids: tree.get_unsafe(&node_id).children().as_slice().iter() }
+    }
+}
+
+impl<'a> Iterator for ChildrenIds<'a> {
+    type Item = &'a NodeId;
+
+    fn next(&mut self) -> Option<&'a NodeId> {
+        self.child_ids.next()
+    }
+}
+
+pub struct PreOrderTraversal<'a, T: 'a> {
+    tree: &'a Tree<T>,
+    node_id: Option<NodeId>,
+    future_data: Vec<&'a NodeId>,
+}
+
+impl<'a, T> IteratorNew<'a, T, PreOrderTraversal<'a, T>> for PreOrderTraversal<'a, T> {
+    fn new(tree: &'a Tree<T>, node_id: NodeId) -> PreOrderTraversal<T> {
+        PreOrderTraversal {
+            tree: tree,
+            node_id: Some(node_id),
+            // over allocating, but all at once instead of re-sizing and re-allocating as we go
+            future_data: Vec::with_capacity(tree.nodes.capacity()),
+        }
+    }
+}
+
+impl<'a, T> Iterator for PreOrderTraversal<'a, T> {
+    type Item = &'a Node<T>;
+
+    fn next(&mut self) -> Option<&'a Node<T>> {
+        if let Some(node_id) = self.node_id.clone() {
+
+            let node = self.tree.get_unsafe(&node_id);
+
+            if !node.children().is_empty() {
+                let mut children_ids = node.children().iter();
+
+                // take the first child and set it as the next one to be processed
+                self.node_id = children_ids.next().cloned();
+
+                // reverse it because we will pop to get the next one later.
+                let mut new_ids: Vec<&NodeId> = children_ids.rev().collect();
+
+                self.future_data.append(&mut new_ids);
+            } else {
+                self.node_id = self.future_data.pop().cloned();
+            }
+
+            return Some(node);
+        }
+        None
+    }
+}
+
+pub struct PostOrderTraversal<'a, T: 'a> {
+    tree: &'a Tree<T>,
+    index: usize,
+    data: Vec<NodeId>,
+}
+
+impl<'a, T> PostOrderTraversal<'a, T> {
+    fn process_nodes<'b>(&mut self, starting_id: NodeId) {
+        let node = self.tree.get_unsafe(&starting_id);
+
+        for child_id in node.children() {
+            self.process_nodes(child_id.clone());
+        }
+
+        self.data.push(starting_id);
+    }
+}
+
+impl<'a, T> IteratorNew<'a, T, PostOrderTraversal<'a, T>> for PostOrderTraversal<'a, T> {
+    fn new(tree: &'a Tree<T>, node_id: NodeId) -> PostOrderTraversal<T> {
+        let mut traversal = PostOrderTraversal {
+            tree: tree,
+            index: 0,
+            // over allocating, but all at once instead of re-sizing and re-allocating as we go
+            data: Vec::with_capacity(tree.nodes.capacity()),
+        };
+
+        traversal.process_nodes(node_id);
```

Comment:
Yeah, good points.  That would probably be the way to go.  I'm probably going to leave it for now and create an issue for later. Thanks for taking a look!